feat: validate `SDConfig.SUPPORTED_LANGUAGES` for usable locales

cfm opened the pull request:
## Status

Ready for review


## Description of Changes

Closes #6366 by determining that a locale is *usable* if it is both (a) *available* in the filesystem and (b) *configured* by the administrator in `SDConfig.SUPPORTED_LANGUAGES`.  Once we've determined which configured locales are actually usable, we:

1. warn if a configured locale is not available;

2. fall back to the hard-coded `FALLBACK_LOCALE` (`en_US`) if `SDConfig.DEFAULT_LOCALE` is not usable; and

3. error out if neither the default nor the fallback locale is usable.


## Testing

### Routine testing

In the development environment:

* [x] Confirm that the language menu works normally.
    * [x] `zh_{Hans,Hant}` are each (separately) listed, selectable, and loadable.

### Configured language loses support

In the development environment:
    
1. First:

    ```sh-session

   $ rm -rf securedrop/translations/de_DE  # or the language of your choice
   $ make dev
   ```
2. [x] Observe the error:

    ```sh-session

    [2022-04-19 22:14:24,290] ERROR in i18n: Configured locales {Locale('de', territory='DE')} are not in the set of usable locales {Locale('es', territory='ES'), Locale('is'), Locale('pt', territory='BR'), Locale('zh', script='Hans'), Locale('ro'), Locale('it', territory='IT'), Locale('en', territory='US'), Locale('ca'), Locale('nl'), Locale('fr', territory='FR'), Locale('el'), Locale('sv'), Locale('ar'), Locale('cs'), Locale('ru'), Locale('hi'), Locale('nb', territory='NO'), Locale('tr'), Locale('zh', script='Hant'), Locale('sk')}

    ```

3. [x] In the Source Interface, observe that **Deutsch** is no longer available in the language menu.

### Default language loses support

In the development environment:
    
1. Apply the patch:

    ```patch

    --- a/securedrop/config.py.example

    +++ b/securedrop/config.py.example

    @@ -82,7 +82,7 @@ DATABASE_ENGINE = 'sqlite'

     DATABASE_FILE = os.path.join(SECUREDROP_DATA_ROOT, 'db.sqlite')

     

     # Which of the available locales should be displayed by default ?

    -DEFAULT_LOCALE = 'en_US'

    +DEFAULT_LOCALE = 'de_DE'

     

     # How long a session is valid before it expires and logs a user out

     SESSION_EXPIRATION_MINUTES = 120

    ```

2. Then:

    ```sh-session

    $ rm -rf securedrop/translations/de_DE # or the language of your choice

    $ make dev

    ```

3. [x] Observe the error:

    ```sh-session

    [2022-04-19 22:14:24,290] ERROR in i18n: Configured locales {Locale('de', territory='DE')} are not in the set of usable locales {Locale('es', territory='ES'), Locale('is'), Locale('pt', territory='BR'), Locale('zh', script='Hans'), Locale('ro'), Locale('it', territory='IT'), Locale('en', territory='US'), Locale('ca'), Locale('nl'), Locale('fr', territory='FR'), Locale('el'), Locale('sv'), Locale('ar'), Locale('cs'), Locale('ru'), Locale('hi'), Locale('nb', territory='NO'), Locale('tr'), Locale('zh', script='Hant'), Locale('sk')}

    ```

4. [x] In the Source Interface, observe that **Deutsch** is no longer available in the language menu.

### No usable fallback locale

In the development environment:
    
1. Apply the patch:

    ```patch

    --- a/Makefile

    +++ b/Makefile

    @@ -171,7 +171,6 @@ securedrop/config.py: ## Generate the test SecureDrop application config.

                     ctx.update(dict((k, {"stdout":v}) for k,v in os.environ.items())); \

                     ctx = open("config.py", "w").write(env.get_template("config.py.example").render(ctx))'

            @echo >> securedrop/config.py

    -       @echo "SUPPORTED_LOCALES = $$(if test -f /opt/venvs/securedrop-app-code/bin/python3; then ./securedrop/i18n_tool.py list-locales --python; else DOCKER_BUILD_VERBOSE=false $(DEVSHELL) ./i18n_tool.py list-locales --python; fi)" | sed 's/\r//' >> securedrop/config.py

            @echo

     

     .PHONY: test-config

    diff --git a/securedrop/config.py.example b/securedrop/config.py.example

    index 92544087a..508e5db66 100644

    --- a/securedrop/config.py.example

    +++ b/securedrop/config.py.example

    @@ -82,7 +82,8 @@ DATABASE_ENGINE = 'sqlite'

     DATABASE_FILE = os.path.join(SECUREDROP_DATA_ROOT, 'db.sqlite')

     

     # Which of the available locales should be displayed by default ?

    -DEFAULT_LOCALE = 'en_US'

    +DEFAULT_LOCALE = 'de_DE'

     

     # How long a session is valid before it expires and logs a user out

     SESSION_EXPIRATION_MINUTES = 120

    +SUPPORTED_LOCALES = ['de_DE']

    ```

2. Then:

    ```sh-session

    $ rm -rf securedrop/config.py
    $ rm -rf securedrop/translations/de_DE # or the language of your choice

    $ make dev

    ```

3. [x] Observe the crash:

    ```

    ValueError: None of the default locales {Locale('en', territory='US'), Locale('de', territory='DE')} is in the set of usable locales set()

    ```


### OSSEC alerts

In the staging environment:

1. On the Monitor Server, `tail -f /var/ossec/logs/alerts/alerts.log`.
2. On the Application Server:

    ```sh-session

    $ sudo rm -rf /var/www/securedrop/translations/de_DE  # or another locale configured in your staging environment

    $ sudo systemctl restart apache2

    ```

3. [x] On the Monitor Server, observe an alert like:

    ```sh-session

    ** Alert 1650408345.42633: mail  - Apache logs

    2022 Apr 19 22:45:45 (sd-staging-app-base-focal) 10.137.0.50->/var/log/apache2/journalist-error.log

    Rule: 400700 (level 7) -> 'Apache application error.'

    [2022-04-19 22:14:24,290] ERROR in i18n: Configured locales {Locale('de', territory='DE')} are not in the set of usable locales {Locale('es', territory='ES'), Locale('is'), Locale('pt', territory='BR'), Locale('zh', script='Hans'), Locale('ro'), Locale('it', territory='IT'), Locale('en', territory='US'), Locale('ca'), Locale('nl'), Locale('fr', territory='FR'), Locale('el'), Locale('sv'), Locale('ar'), Locale('cs'), Locale('ru'), Locale('hi'), Locale('nb', territory='NO'), Locale('tr'), Locale('zh', script='Hant'), Locale('sk')}

   ```


## Deployment

This is a *disruptive* change for an instance X if:

1. we remove a language L for lack of translation coverage; and
2. L is configured as X's `DEFAULT_LANGUAGE`.

This is a *breaking* change for X if, in addition:

3. the fallback locale `en_US` has been *removed* from X's `SUPPORTED_LANGUAGES`.

A report of instances' `supported_languages` from the Directory should tell us whether any instances will need proactive outreach about either of these scenarios.  Otherwise, we should make sure to warn about them in the release notes.


## Checklist

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

Choose one of the following:

- I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- I would appreciate help with the documentation
- [x] These changes do not require documentation

cfm committed:
feat: validate SDConfig.SUPPORTED_LANGUAGES for *usable* locales

A locale is considered usable if it is both (a) available in the
filesystem and (b) configured by the administrator in
SDConfig.SUPPORTED_LANGUAGES.  Once we've determined which configured
locales are actually usable, we:

1. warn if a configured locale is not available;

2. fall back to the hard-coded FALLBACK_LOCALE ("en_US") if
   SDConfig.DEFAULT_LOCALE is not usable; and

3. error out if neither the default nor the fallback locale is usable.

cfm committed:
feat: map_locale_display_names(): skip (hide) unusable locales

We can't iterate directly over USABLE_LOCALES: Set(Locale) because
Locales aren't sortable.  Instead, we parse config.SUPPORTED_LOCALES in
order and skip those that aren't in USABLE_LOCALES.

legoktm commented on securedrop/i18n.py:
```diff
@@ -126,32 +126,39 @@ def configure_babel(config: SDConfig, app: Flask) -> Babel:
 
 
+def parse_locale_set(codes: List[str]) -> Set[Locale]:
+    return set(map(lambda code: Locale.parse(code), codes))
```
nit: I'm not a huge fan of `map` in Python, personally I'd find `return {Locale.parse(code) for code in codes}` to be easier to read.

cfm replied:
Much nicer, I agree.  Done in 9c119506a9f9a67929491ad79011bcb323cec39e.

legoktm commented on securedrop/i18n.py:
```diff
@@ -126,32 +126,39 @@ def configure_babel(config: SDConfig, app: Flask) -> Babel:
+    defaults = parse_locale_set([config.DEFAULT_LOCALE, FALLBACK_LOCALE])
+    if not defaults & usable:
+        raise ValueError(
+            f'None of the default locales {defaults} is in the set of usable locales {usable}'
```
s/is/are/ I think

cfm replied:
Of the changes you've requested, none is more vexing than this one.  ;-)

I had to look this up!  Both are grammatically correct, I think.  But "is" is stuffy, and "are" is more consistent with the logic of `if not defaults & usable` anyway.  Changed in 9c119506a9f9a67929491ad79011bcb323cec39e.

legoktm commented on securedrop/i18n.py:
```diff
@@ -126,32 +126,39 @@ def configure_babel(config: SDConfig, app: Flask) -> Babel:
-                    parsed, translations
-                )
-            )
+    global USABLE_LOCALES
```
I guess we already do this for LOCALES, but I think continuing the pattern should be accompanied by a TODO that manipulating/relying on global state is not desirable.

cfm replied:
Very fair.  Filed in #6420 and flagged in 9c119506a9f9a67929491ad79011bcb323cec39e.

legoktm commented on securedrop/i18n.py:
```diff
@@ -126,32 +126,39 @@ def configure_babel(config: SDConfig, app: Flask) -> Babel:
 
 
 LOCALES = collections.OrderedDict()  # type: collections.OrderedDict[str, RequestLocaleInfo]
+USABLE_LOCALES = set()  # type: Set[Locale]
```
I mostly understand why you typed this as `Set[Locale]` but below, you have to call `Locale.parse(...)` every time you check what's in the set, so wouldn't it be easier to have this be `Set[str]`? or should SUPPORTED_LOCALES actually be a `List[Locale]`??

cfm replied:
Good question.  I've documented this better inline in 9c119506a9f9a67929491ad79011bcb323cec39e.  But in brief:  `SUPPORTED_LOCALES` comes from and is typed as `List[str]` by `sdconfig.SDConfig`, where it's validated only against the `List[str]` corresponding to the `securedrop/translations/` directory available to `securedrop-admin sdconfig`.  Of course, that's different from the `securedrop/translations/` directory available to the Flask apps, which is what we get from [`babel.list_translations() -> List[Locale]`](https://github.com/python-babel/flask-babel/blob/ec7ae9ed2e22c7aebd4e732c1c3dc6d45fe8db76/flask_babel/__init__.py#L136-L139), from which `USABLE_LOCALES` is derived.

So, while I'm certainly open to refactoring suggestions, this is exactly one of the places where this mismatch is both unavoidable and meaningful to bridge.  Does this make sense?

legoktm replied:
Yeah, thanks for the clarification. One idea I'm having now is to have a dynamic property on SDConfig like:

```python3
@property
def REAL_SUPPORTED_LOCALES() -> List[Locale]:
    return [Locale.parse(x) for x in self.SUPPORTED_LOCALES]
```

or something. BUT that's really getting out of scope here, I'll try to file an issue in the next few days with some other ideas/complaints about SDConfig.

legoktm commented on securedrop/tests/test_i18n.py:
```diff
@@ -32,13 +32,17 @@
 from werkzeug.datastructures import Headers
 
 
+NEVER_LOCALE = 'eo'  # Esperanto
```
Just wait until one of the sites listed at https://esperanto.net/en/news-in-esperanto/ sets up a SecureDrop ;)

legoktm commented:
Still walking through the test plan (more feedback to come shortly...), but when I switched my locale to de_DE, deleted it and restarted the dev container, I got a stack trace of:
```
Traceback (most recent call last):
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 2091, in __call__
    return self.wsgi_app(environ, start_response)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 2076, in wsgi_app
    response = self.handle_exception(e)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/home/user/github/freedomofpress/securedrop/securedrop/source_app/main.py", line 38, in index
    return render_template('index.html')
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/templating.py", line 147, in render_template
    return _render(
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/templating.py", line 128, in _render
    rv = template.render(context)
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/jinja2/environment.py", line 1304, in render
    self.environment.handle_exception()
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/jinja2/environment.py", line 925, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/home/user/github/freedomofpress/securedrop/securedrop/source_templates/index.html", line 66, in top-level template code
    {% include 'locales.html' %}
  File "/home/user/github/freedomofpress/securedrop/securedrop/source_templates/locales.html", line 4, in top-level template code
    {{ gettext('Selected language')}}: {{ g.locales[g.localeinfo.id].display_name }}
  File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/jinja2/environment.py", line 474, in getattr
    return getattr(obj, attribute)
jinja2.exceptions.UndefinedError: 'collections.OrderedDict object' has no attribute 'de_DE'
```

Resetting my Tor Browser identity/session fixed it so I don't think it's a problem in practice for production installs, but flagging it as it seems to reveal another place not using/checking USABLE_LOCALES first.

legoktm commented on securedrop/i18n.py:
```diff
@@ -165,11 +172,17 @@ def map_locale_display_names(config: SDConfig) -> None:
 
     locale_map = collections.OrderedDict()
     for l in sorted(config.SUPPORTED_LOCALES):
+        if Locale.parse(l) not in USABLE_LOCALES:
```
Seems weird to do this twice. Can't we combine the two for loops pretty trivially?

```python3
    for l in sorted(config.SUPPORTED_LOCALES):
        if Locale.parse(l) not in USABLE_LOCALES:
            continue

        locale = RequestLocaleInfo(l)
        language_locale_counts[locale.language] += 1
        if language_locale_counts[locale.language] > 1:
            locale.use_display_name = True
            locale_map[str(locale)] = locale
```

Or a set based approach:
```python3
    seen: Set[str] = set()
    locale_map = collections.OrderedDict()
    for l in sorted(config.SUPPORTED_LOCALES):
        if Locale.parse(l) not in USABLE_LOCALES:
            continue

        locale = RequestLocaleInfo(l)
        if locale.language in seen:
            # This is the second time it's appeared
            locale.use_display_name = True
            locale_map[str(locale)] = locale
        else:
            seen.add(locale.language)
```

cfm replied:
Absolutely; I was going for the minimal intervention here.  :-) Refactored (with corrected logic) per your second approach in bc5e3661babc4ff8af7fb8fe18b542daeef1b984; test plan updated accordingly.

legoktm reviewed (changes requested):
dev test plan passed, minus the one stacktrace I posted. I didn't go onto the staging part yet, will do that in the next round of review.

cfm commented:
Thanks, @legoktm!  I'll try to turn this around this afternoon (Pacific) in the hope of getting it under the feature-freeze wire for v2.4.0.

cfm committed:
refactor(map_locale_display_names): use set rather than looping twice to disambiguate SUPPORTED_LOCALES

thread: https://github.com/freedomofpress/securedrop/pull/6406#discussion_r863117986

cfm committed:
fix(get_locale): don't assume the session-provided locale is usable

This is a safe assumption except in the corner case where (e.g.):

1. A user begins a session and selects language L.

2. The Application Server upgrades to a version of securedrop-app-code
   that removes support for language L.

3. The user resumes their session, which still has language L selected.

But, since this case has come up in testing[1], best to check for it.

[1]: https://github.com/freedomofpress/securedrop/pull/6406#issuecomment-1115257623

cfm committed:
refactor(get_locale): resolve locale from single list of preferences

As of 9c11950, get_locale() handled user/agent-provided locale
preferences and deferred to resolve_fallback_locale() if none
could be satisfied.  Here we build a single list of locales,
beginning with user/agent preferences and ending with the server-side
{DEFAULT_FALLBACK}_LOCALEs, and call babel.core.negotiate_locale()
once to negotiate one (if possible).

cfm commented:
@legoktm in <https://github.com/freedomofpress/securedrop/pull/6406#issuecomment-1115257623>:

> Still walking through the test plan (more feedback to come shortly...), but when I switched my locale to de_DE, deleted it and restarted the dev container, I got a stack trace of:
> 
> [...]
>  
> Resetting my Tor Browser identity/session fixed it so I don't think it's a problem in practice for production installs, but flagging it as it seems to reveal another place not using/checking USABLE_LOCALES first.

Thanks for catching this failure-mode.  I'm not able to reproduce it, but I've added what I think is the necessary defense-in-depth check in eb2a042f59dc353e34ca0444029d154d28b0a7d3.  If you're still able to reproduce with this change, and think it worth addressing, could you specify a more-detailed reproduction in terms of the "delete `de_DE`" and "restart container" steps, to make sure I'm testing what you're seeing?

This change prompted the further refactoring proposed in eb2a042f59dc353e34ca0444029d154d28b0a7d3, to consolidate the locale-negotiation logic in `get_locale()`.  Let me know what you think, or feel free to rebase out for the sake of an expedient review and merge for v2.4.0.

legoktm commented:
Using your latest set of changes I can no longer reproduce it, so I assume it's fixed. Going through the rest of the test plan now...

legoktm reviewed (approved):
Tested on both dev and staging, LGTM!
